### PR TITLE
Implemented sub_69A5F0

### PR DESCRIPTION
**Pull request by LRFLEW**
This is my first attempt at decompilation, so if somebody wants to double-check my work, please do so.

### CODE REVIEW

File: src/peep/peep.c [hunk 1]
```diff
@@ -7131,70 +7131,106 @@ uint8 sub_69A60A(rct_peep* peep){
...
+			}
+		}
+	}
+
+    // May be redundant, but shouldn't be removed until it's confirmed that sub_69A997 doesn't modify them
+	x_goal = RCT2_GLOBAL(RCT2_ADDRESS_PEEP_PATHFINDING_GOAL_X, sint16) / 32;
+	y_goal = RCT2_GLOBAL(RCT2_ADDRESS_PEEP_PATHFINDING_GOAL_Y, sint16) / 32;
+	z_goal = RCT2_GLOBAL(RCT2_ADDRESS_PEEP_PATHFINDING_GOAL_Z, uint8);
+
+	if (peep->pathfind_goal.direction > 3 || peep->pathfind_goal.x != x_goal || peep->pathfind_goal.y != y_goal || peep->pathfind_goal.z != z_goal) {
```
Comment (duncanspumpkin) [inline src/peep/peep.c]: Make this go onto 4 lines then it will be easier to read

File: src/peep/peep.c [hunk 2]
```diff
@@ -7131,70 +7131,106 @@ uint8 sub_69A60A(rct_peep* peep){
...
+		peep->pathfind_goal.x = x_goal;
+		peep->pathfind_goal.y = y_goal;
+		peep->pathfind_goal.z = z_goal;
+		peep->pathfind_goal.direction = 0;
+		memset(peep->pathfind_history, 0xFF, sizeof(peep->pathfind_history)); // Clear pathfinding history
+	}
+
+	for (int i = 0; i < 4; ++i) {
+		if (peep->pathfind_history[i].x == x / 32 && peep->pathfind_history[i].y == y / 32 && peep->pathfind_history[i].z == z) {
```
Comment (duncanspumpkin) [inline src/peep/peep.c]: Same for this

File: src/peep/peep.h [hunk 3]
```diff
@@ -507,8 +507,8 @@ typedef struct {
 	};
 	uint8 photo1_ride_ref;			// 0xC7
 	uint32 flags;					// 0xC8
-	rct_xyzd8 var_CC;
-	rct_xyzd8 var_D0[4];
+	rct_xyzd8 pathfind_goal;
```
Comment (duncanspumpkin) [inline src/peep/peep.h]: Can you add the offsets as a comment

File: src/peep/peep.c [hunk 4]
```diff
@@ -8401,10 +8437,10 @@ static bool peep_should_use_cash_machine(rct_peep *peep, int rideIndex)
  */
 static void sub_69A98C(rct_peep *peep)
```
Comment (duncanspumpkin) [inline src/peep/peep.c]: Relabel as 'reset_pathfind_goal'

Comment (LRFLEW) [inline src/peep/peep.c]: 'reset_pathfind_goal' isn't really what this does. I'm thinking `pathfind_choose_direction` would be more descriptive.

File: src/peep/peep.c [hunk 5]
```diff
@@ -7131,70 +7131,115 @@ uint8 sub_69A60A(rct_peep* peep){
...
-	//	for (; index < 4; ++index){
-	//		if (peep->var_D0[index].x == x &&
-	//			peep->var_D0[index].y == y &&
-	//			peep->var_D0[index].z == z){
-	//			edges = peep->var_D0[index].direction & 0xF;
-	//			break;
-	//		}
-	//	}
-	//}
+static int pathfind_choose_direction(sint16 x, sint16 y, sint16 z, rct_peep *peep, rct_map_element *map_element) {
```
Comment (duncanspumpkin) [inline src/peep/peep.c]: This function will only be used by guests so could you prepend `guest_` onto the function name.

File: src/peep/peep.c [hunk 6]
```diff
@@ -7131,70 +7131,115 @@ uint8 sub_69A60A(rct_peep* peep){
 
 /**
  *
+ *  rct2: 0x0069A997
+ */
+static int sub_69A997(sint16 x, sint16 y, uint8 z, int direction, rct_map_element *dest_map_element, int test_edge) {
+	int eax = x, ebx = direction, ecx = y, edx = z, esi = (int) dest_map_element, edi = 0xFFFF, ebp = test_edge;
+	RCT2_CALLFUNC_X(0x0069A997, &eax, &ebx, &ecx, &edx, &esi, &edi, &ebp);
```
Comment (duncanspumpkin) [inline src/peep/peep.c]: `direction`, `dest_map_element` is not an input into this function. `dh` and `dl` (`z`) are separate variables to this function.

File: src/peep/peep.c [hunk 7]
```diff
@@ -8399,12 +8444,12 @@ static bool peep_should_use_cash_machine(rct_peep *peep, int rideIndex)
  *
  *  rct2: 0x0069A98C
  */
-static void sub_69A98C(rct_peep *peep)
+static void reset_pathfind_goal(rct_peep *peep)
```
Comment (duncanspumpkin) [inline src/peep/peep.c]: This function is used by both guests and staff so should really be prepended `peep_`.

**Comment by duncanspumpkin** [comment]
Sorry I forgot most of the functions in peep.c have the prepended names could you make those changes and then this should be ready for merging.

File: src/peep/peep.c (hunk 6 above)
Comment (LRFLEW) [inline src/peep/peep.c]: I wasn't sure what got used by the other function, so I played it safe. If you know that neither are used, then I'll remove them. 

You also said that dh was a separate variable. Does it have a value I should be initializing it to? As far as I could tell, it was just being assigned `z` from the caller function.

Comment (duncanspumpkin) [inline src/peep/peep.c]: `dh` appears to be set to 0 by the calling function `pathfind_choose_direction` but then sub_69A997 calls itself recursively with different values of `dh`.

Comment (LRFLEW) [inline src/peep/peep.c]: >sub_69A997 calls itself recursively with different values of `dh`.

Oh boy. I think I'm going to tackle that one next, but it sounds messy :P

Comment (duncanspumpkin) [inline src/peep/peep.c]: On the plus side it looks quite short.

**Comment by LRFLEW** [comment]
I've made the improvements requested, and this is ready for merging. I realized what this code actually means for the peep AI, and it seems to explain a lot on how peep jams occur.

**Pull request merged**